### ISSUE TITLE
segfault when "viewing" closed mmap object

### ISSUE THREAD

**Issue by ahaldane** [numpy/numpy#9537]
This crashes in python2 and python3 with a segfault:
```python
import mmap

# create test file
a = arange(10., dtype='f4')
a.tofile('test')

# opens it as a memmap
with open('test', 'r+b') as f:
      mm = mmap.mmap(f.fileno(),0)

#view the mmap, then close it
x = np.frombuffer(mm, dtype='f4')
mm.close()
print(x)  # segfault
```

In `frombuffer` mumpy uses the PEP3118 interface to the mmap to get a pointer to the data, so it seems we are using the interface incorrectly somewhere.

It seems to me that the "correct" behavior may be for it to be impossible to close the memmap while pointers to it exist; this is the behavior for `memoryview`s of mmaps. That is, the line `mm.close()` shoud raise an error `BufferError: cannot close exported pointers`, as in currently happens in python3:
```python
>>> import mmap
>>> with open('test', 'r+b') as f:
>>>     mm = mmap.mmap(f.fileno(),0)
>>> mv = memoryview(mm)
>>> mm.close()
BufferError: cannot close exported pointers exist
```

however note also that in python2 we cannot create the memoryview (`mv`) because of an error `TypeError: cannot make memory view because object does not have the buffer interface`.

Note how [mmap_buffer_getbuf (link)](https://github.com/python/cpython/blob/b879fe82e7e5c3f7673c9a7fa4aad42bd05445d8/Modules/mmapmodule.c#L727) in the mmap module takes care to keep track of any buffer access to the mmap (the `exports` counter). Evidently we are somehow bypassing that in numpy.

**Comment by ahaldane** [comment]
This was reported on the mailing list in [this message](https://mail.python.org/pipermail/numpy-discussion/2017-August/077105.html)

**Comment by orkolorko** [comment]
The bug was not found by me, but by Maurizio Monge maurizio.monge at im.ufrj.br; I just used it to check the behaviour of numpy in a specific case.

**Comment by pv** [comment]
Maybe it ends up using the legacy buffer API, which iirc has no such
safety guarantees?

**Comment by jakirkham** [comment]
There were some fixes for buffers recently in PR ( https://github.com/numpy/numpy/pull/10740 ) included in NumPy 1.15.0. Does this issue persist with those changes?

**Comment by eric-wieser** [comment]
No errors on python 3 for me in 1.17.x

**Comment by seberg** [comment]
Does segfault for me, maybe windows is different? I somewhat thought I saw a warning/error here on closing the memmap even though it still has references.

**Comment by eric-wieser** [comment]
My mistake, failed to copy the `print` line when I tested this. Fails on windows for me.

**Comment by devinrsmith** [comment]
I've run into similar issues when trying to use `np.frombuffer` against our own objects that implement the buffer protocol. I think the issue is that numpy mis-uses the buffer api - it gets and immediately releases the buffer instead of releasing the buffer during the the returned arrays destructor.

https://github.com/numpy/numpy/blob/v1.18.3/numpy/core/src/multiarray/ctors.c#L3817-L3823

**Comment by seberg** [comment]
To be clear, this comes from a time when the code would have been perfectly correct, because `view->obj` was always defined as the exporting object, and `PyBuffer_Release` was not able to do anything more than decreasing the refcount of `obj`.

NumPy itself doesn't even *use* the "free" function right now, because it leads to annoying side effects with the `PyArg_Parse*` functions.

So, as long as the data pointed to by your object remains available for the lifetime of that object, you are actually fine. Otherwise, it seems we would have to do the internal reference counting dance that `memoryview` has to do. I do not know how tricky that is, but I am not sure I am willing to do that. So the only thing I can think of is: Wrap all "buffers" directly in memoryviews. I wonder if that would be better to begin with... if holding on to the buffer is important for the memory to remain valid, why not promote it to a python object, they already do memory management?

But of course, in NumPy we thus use `sys.refcount` in one or two places, to warn users that they are doing invalid things. With mmap+memoryview, you actually get a real warning also on PyPy, since it does its own reference counting...

Long story short, I did not realize that memoryviews had this ingrained so much, and `mmap` actually *uses* it semi-prominently when I opened https://bugs.python.org/issue39471 to at least clarify things. I am not sure we should even consider doing that internal refcounting dance... But luckily, we now have only python 3, so we could just wrap everything in a `memoryview` and use that. Of course after that we may have to brace for impact of users expecting `arr.base` not to be a `memoryview` suddenly?

**Comment by devinrsmith** [comment]
Thanks for chiming in - I certainly don't know the intricate implementation details of numpy, but I think the assumption that "the object remains available for the lifetime of that object" might be a bad assumption. At the very least, it needs to be warned and documented.

For example, this seems like it should be valid:

```python
def to_numpy(x: SomeCustomInt32BufferProtocol):
  return np.frombuffer(x, dtype=np.int32)

y = to_numpy(SomeCustomInt32BufferProtocol(...))
```

Due to the current implementation of `np.frombuffer` though, there is no way for the custom impl to know that anything is still referencing its underlying buffer.

I think due to the inherent nature of the buffer protocol, libraries/consumers that call `PyObject_GetBuffer` are expected to manage their calls back into `PyBuffer_Release` with care.

**Comment by eric-wieser** [comment]
@seberg, I'm afraid I never quite followed the conclusion of your one-sided discussion on that issue. Was it ever clear what the intended semantics were?

**Comment by devinrsmith** [comment]
I might be mis-interpreting your issue https://bugs.python.org/issue39471, but it seems like the testbuffer impl *doesn't* allocate during the getbufferproc stage:

https://github.com/python/cpython/blob/v3.8.2/Modules/_testbuffer.c#L1527-L1529

That said, I think it's completely legitimate for getbufferproc to allocate instead of returning "views" - it should be an implementation detail that shouldn't matter to a generic consumer (the generic consumer in this case being `np.frombuffer`). Of course, as a specific consumer, using a specific buffer protocol library, I should be aware about the expectations of the implementation. IE, mmap and np.array implement the buffer protocol with efficient "view" semantics.

As a hypothetical - I could implement a getbufferproc that returns new random bytes for each call.

**Comment by seberg** [comment]
Well, the conclusion is that it annoys me, but in the end I have to say that the way NumPy does it probably must be considered wrong at this time...

I expect NumPy was doing the right thing until the buffer protocol got revised and `memoryviews` came to be with Python 3. At which point the semantics were refined and the assumption that holding on to `obj` is good enough is just not good enough anymore.
A change that never made it into NumPy, and for which there was never even an attempt to get right (there was an attempt to get my interpretation right, which is probably why I had that impression).

Now `mmap` assumes the data must remain valid (only) until `PyBuffer_Release` is called and `memoryview` has all the bells and whistles necessary to get it right...
So @eric-wieser yes, it was probably just me having the misconception, since NumPy must probably be considered broken as soon as the buffer PEP was accepted.

As far as I can tell the only way to fix this will be implement `frombuffer` as:
```
def frombuffer(x):
     if isinstance(x, np.ndarray):
         return x.view()  # effectively
     return current_frombuffer(memoryview(x))
```
`arr.base` will almost always be a `memoryview` object after that. As far as I can tell, any other solution must include adding new fields to the `ndarray` object/struct since we would have to implement reference counting to some internal buffer objects...

**Comment by seberg** [comment]
@devinrsmith yes you are right, although I could argue that if you have such complex use, we should maybe be passing around `memoryview` objects to begin with.
That would simplify most of the memory management. Since only `memoryview` has to worry about the complexities of setting up a secondary reference counting system to the underlying buffer interface... NumPy is happy to *not* reallocate arrays after all (at least for all practical purposes).

**Comment by devinrsmith** [comment]
@seberg thanks again for your insights. The buffer protocol semantics aren't the easiest to grok, especially when python transitioned from the old api to the new api, I can see how things got muddy.

I think it's probably good general advice that if you are implementing the buffer protocol, and you are consuming an existing buffer, it makes sense to use a `memoryview` to simplify your bookkeeping. In my exact case though, I'm not trying implementing the buffer protocol from an existing buffer protocol object (in my domain, something like `my_lib.frombuffer(buf)` doesn't even make sense).

It's good to know though that wrapping w/ `memoryview` is a workaround to achieve the expected buffer protocol semantics.

**Comment by eric-wieser** [comment]
> As far as I can tell the only way to fix this will be implement `frombuffer` as:
> 
> ```
> def frombuffer(x):
>      if isinstance(x, np.ndarray):
>          return x.view()  # effectively
>      return current_frombuffer(memoryview(x))
> ```

Presumably it actually becomes
```python
def frombuffer(x):
     if has_buffer_release(x):
         x = memoryview(x)
     return current_frombuffer(x)
```

**Comment by seberg** [comment]
@eric-wieser true, that means you have to look at the slots of the exporting type. Which however is not a problem. Although I have no idea how much difference that will actually make in practice.

But yes, checking `Py_TYPE(obj)->tp_as_buffer.bf_releasebuffer != NULL` should work and be a bit more general than just NumPy arrays.

We will, and maybe just should, just live with changing the `.base` attribute. This has been dangling for years, so if nobody is in a rush, maybe something to do after the 1.19 release so that it lurks in master a bit longer?

**Comment by devinrsmith** [comment]
Checking `bf_releasebuffer` *seems* like it should work (https://github.com/python/cpython/blob/v3.8.2/Objects/abstract.c#L690-L692), and I don't have any concrete examples where it wouldn't work, but doing that *may* be making subtle other assumptions about the impl that doesn't hold up like it does with the ndarray impl.

Either way - it's probably good to get some input from a cpython buffer protocol maintainer.

**Comment by seberg** [comment]
The only assumption that it has, is that `buffer->obj` is going to be the exporting object and not some new object created during the export. I think that is fine, if you want to be pedantic, I guess we could check that.
Of course the only reason to do that is to safe a few cpu cycles creating a memorymap and retain the `arr.base` attribute. So it may not be worth the trouble.
CPython currently uses (has to use) that assumption also during argument parsing to retain backward compatibility (the reason why NumPy does not use `releasebuffer`, which is seriously annoying).

**Comment by eric-wieser** [comment]
> CPython currently uses (has to use) that assumption also during argument parsing to retain backward compatibility 

In what way is that a compatibility choice? Presumably the issue is that `ParseTuple...` and friends do not have a code that means "get me a pointer _and_ and object to manage its lifetime with".

**Comment by eric-wieser** [comment]
> Either way - it's probably good to get some input from a cpython buffer protocol maintainer.

@skrah and @abalkin might be your best bet there, at least based on the reviewers assigned to python/cpython#5576 and python/cpython#5561

**Comment by seberg** [comment]
Because `PyArg_ParseTuple("%z")` (or whatever), can keep working with NumPy arrays because NumPy arrays do not implement `releasebuffer` and thus it can assume that the pointer is valid for the whole lifetime of the exporting object. So I think it is "compatibility" with the old buffer-protocol where the lifetime assumption was correct. And the reason why NumPy is stuck with it currently.

**Comment by eric-wieser** [comment]
The compatibility choice is _ours_ though right, not cpython's? Cpython's behavior is required to be correct.

**Comment by seberg** [comment]
@eric-wieser I guess I have a different angle. Implicitly CPython choose to say that the old-buffer protocol definition (lifetime is bound to exporting object) *is* valid. Otherwise they would have to deprecate and remove that special case path, leaving NumPy no choice (or no gain in sticking with the old implementation).
Of course, in all likelyhood nobody ever seriously tried to think about deprecating it in CPython, and simply put in the backward compatibility skim to close an issue or two.

**Comment by seberg** [comment]
So maybe its different. The choice is ours, but CPython gives us less choice to move to the new buffer-protocol definition.
And maybe the solution is actually to just screw it, and release NumPy with a flag `NPY_USE_RELEASE_BUFFER=0`. Keep at at 0 for a while in release versions, put it to 1 in pre-releases and wait out until the shouting dies down...

**Comment by eric-wieser** [comment]
> . Implicitly CPython choose to say that the old-buffer protocol definition (lifetime is bound to exporting object) _is_ valid

My reading of this is different. CPython has concluded that if your buffer has no release process, then one of the following is true:
* The buffer doesn't need releasing, as its lifetime is bound to the object
* The buffer doesn't need releasing, as its lifetime is global

In essence, the reason goes "I cannot release this buffer, therefore no one can, and I can guarantee it will stick around as long as the pointer I return".

**Comment by seberg** [comment]
Yes, that is correct of course. But it doesn't help with the fact that it puts NumPy in a position where transitioning to the new meaning is tricky. Even though it would help us *a lot* (if we assume `realloc's` do not happen at least – if we do not assume that, it would still help us a lot, but we would have to do more shenanigans, much less than now though).

It seems also a bit annoying that the ownership tracking has to go through `memoryview` objects in this way. I admit, its probably just that something about that setup rubs me the wrong way.

But yes, you are right. There is nothing wrong as such, its just all not ideal and potentially Python could *help* us, since the argument-parsing is the main/only reason I have ever seen for NumPy to remain stuck with the old protocol/interpretation.

**Comment by pv** [comment]
FWIW, Numpy is already using memoryviews for the life cycle management in other places:

https://github.com/numpy/numpy/blob/28895456f147bc6e78f9366b75d1897bbc7ac3fa/numpy/core/src/multiarray/ctors.c#L1532-L1537

This was done, as the `Py_buffer` has to be stored somewhere and it was simplest to let a separate base object manage it. Frombuffer was not touched at the time, as it was using the old unsafe `PyObject_AsReadBuffer/AsWriteBuffer` API that still worked. But maybe it should have been changed back then.
